library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java

```diff
@@ -433,6 +433,13 @@ public boolean onPlaylistError(Uri playlistUrl, long blacklistDurationMs) {
     return chunkIterators;
   }
 
+  public int getPreferredQueueSize(long playbackPositionUs, List<? extends MediaChunk> queue) {
+    if (fatalError != null || trackSelection.length() < 2) {
+      return queue.size();
+    }
+    return trackSelection.evaluateQueueSize(playbackPositionUs, queue);
+  }
+
```

Comment:
I'm looking at actually implementing `ChunkSource` interface.  This looks mostly possible for HLS library except that the `ChunkHolder` is a final class....  Making it an interface and allowing `HlsChunkSource` to implement it looks like the best path.

One of the reasons I'm putting this pull up early is to get feed back on the history and desired architecture direction

Reply:
Got it.  Yes the code is similar.  Main difference is how HLS manages lists of `SampleStreams`.